File: java/src/test/java/javamatch/engine/OrderBookTest.java
package javamatch.engine;

import org.junit.jupiter.api.Test;

import java.util.List;

import static javamatch.engine.EngineAsserts.assertEvents;
import static org.junit.jupiter.api.Assertions.assertEquals;

class OrderBookTest {
    @Test
    void rejectsInvalidOrders() {
        OrderBook b = new OrderBook();
        assertEvents(b.newLimitOrder(new NewOrderCmd(7, 1, Side.BUY, 10, 0)),
            List.of(new Event(EventType.REJECTED).clientOrderId(7).owner(1).reason(RejectReason.BAD_QTY)));
        assertEvents(b.newLimitOrder(new NewOrderCmd(8, 1, Side.BUY, -5, 10)),
            List.of(new Event(EventType.REJECTED).clientOrderId(8).owner(1).reason(RejectReason.BAD_PRICE)));
    }

    @Test
    void restingOrderAcceptedWithBookUpdate() {
        OrderBook b = new OrderBook();
        assertEvents(b.newLimitOrder(new NewOrderCmd(7, 1, Side.BUY, 100, 30)), List.of(
            new Event(EventType.ACCEPTED).orderId(1).clientOrderId(7).owner(1).side(Side.BUY).price(100).qty(30),
            new Event(EventType.BOOK_UPDATE).side(Side.BUY).price(100).aggregateQty(30)));
    }

    @Test
    void orderIdsIncrease() {
        OrderBook b = new OrderBook();
        List<Event> first = b.newLimitOrder(new NewOrderCmd(1, 1, Side.BUY, 100, 1));
        List<Event> second = b.newLimitOrder(new NewOrderCmd(2, 1, Side.SELL, 200, 1));
        assertEquals(1, first.get(0).orderId);
        assertEquals(2, second.get(0).orderId);
    }

    @Test
    void sameLevelAggregates() {
        OrderBook b = new OrderBook();
        b.newLimitOrder(new NewOrderCmd(1, 1, Side.BUY, 100, 30));
        assertEvents(b.newLimitOrder(new NewOrderCmd(2, 2, Side.BUY, 100, 20)), List.of(
            new Event(EventType.ACCEPTED).orderId(2).clientOrderId(2).owner(2).side(Side.BUY).price(100).qty(20),
            new Event(EventType.BOOK_UPDATE).side(Side.BUY).price(100).aggregateQty(50)));
    }
}

File: go/engine/order_book_test.go
package engine

import (
	"reflect"
	"testing"
)

func assertEvents(t *testing.T, got, want []Event) {
	t.Helper()
	if !reflect.DeepEqual(got, want) {
		t.Fatalf("events mismatch\ngot:  %+v\nwant: %+v", got, want)
	}
}

func TestRejectsInvalidOrders(t *testing.T) {
	b := NewOrderBook()
	assertEvents(t, b.NewLimitOrder(NewOrderCmd{ClientOrderId: 7, Owner: 1, Side: Buy, Price: 10, Qty: 0}),
		[]Event{{Type: EvRejected, ClientOrderId: 7, Owner: 1, Reason: ReasonBadQty}})
	assertEvents(t, b.NewLimitOrder(NewOrderCmd{ClientOrderId: 8, Owner: 1, Side: Buy, Price: -5, Qty: 10}),
		[]Event{{Type: EvRejected, ClientOrderId: 8, Owner: 1, Reason: ReasonBadPrice}})
}

func TestRestingOrderAcceptedWithBookUpdate(t *testing.T) {
	b := NewOrderBook()
	got := b.NewLimitOrder(NewOrderCmd{ClientOrderId: 7, Owner: 1, Side: Buy, Price: 100, Qty: 30})
	assertEvents(t, got, []Event{
		{Type: EvAccepted, OrderId: 1, ClientOrderId: 7, Owner: 1, Side: Buy, Price: 100, Qty: 30},
		{Type: EvBookUpdate, Side: Buy, Price: 100, AggregateQty: 30},
	})
}

func TestOrderIdsIncrease(t *testing.T) {
	b := NewOrderBook()
	first := b.NewLimitOrder(NewOrderCmd{ClientOrderId: 1, Owner: 1, Side: Buy, Price: 100, Qty: 1})
	second := b.NewLimitOrder(NewOrderCmd{ClientOrderId: 2, Owner: 1, Side: Sell, Price: 200, Qty: 1})
	if first[0].OrderId != 1 || second[0].OrderId != 2 {
		t.Fatalf("expected ids 1,2 got %d,%d", first[0].OrderId, second[0].OrderId)
	}
}

func TestSameLevelAggregates(t *testing.T) {
	b := NewOrderBook()
	b.NewLimitOrder(NewOrderCmd{ClientOrderId: 1, Owner: 1, Side: Buy, Price: 100, Qty: 30})
	got := b.NewLimitOrder(NewOrderCmd{ClientOrderId: 2, Owner: 2, Side: Buy, Price: 100, Qty: 20})
	assertEvents(t, got, []Event{
		{Type: EvAccepted, OrderId: 2, ClientOrderId: 2, Owner: 2, Side: Buy, Price: 100, Qty: 20},
		{Type: EvBookUpdate, Side: Buy, Price: 100, AggregateQty: 50},
	})
}
